### ISSUE TITLE
set_time_limit exception

### ISSUE THREAD

**Issue by Herokid** [yiisoft/yii2#16531]
Hi,
This line fail when is not possible change max_execution_time (secure mode on shared hosting).

https://github.com/yiisoft/yii2/blob/feae9be281cd897c2eb5d5f9790171bde2ca75d7/framework/web/Response.php#L419

this is the Exception:
`2018-07-17 17:49:57 [79.152.50.37][1][5a0d110aa5700ad29f45e00b218e02a6][error][yii\base\ErrorException:2] yii\base\ErrorException: set_time_limit() has been disabled for security reasons in /vendor/yiisoft/yii2/web/Response.php:419 Stack trace: #0 /vendor/craftcms/cms/src/web/ErrorHandler.php(84): yii\base\ErrorHandler->handleError(2, 'set_time_limit(...', '/usr/home/affor...', 419) #1 [internal function]: craft\web\ErrorHandler->handleError(2, 'set_time_limit(...', '/usr/home/affor...', 419, Array) #2 /vendor/yiisoft/yii2/web/Response.php(419): set_time_limit(0) #3 /vendor/yiisoft/yii2/web/Response.php(340): yii\web\Response->sendContent() #4 /vendor/yiisoft/yii2/base/Application.php(392): yii\web\Response->send() #5 (...)/web/index.php(21): yii\base\Application->run() #6 {main} 2018-07-17 17:49:56 [79.152.50.37][1][5a0d110aa5700ad29f45e00b218e02a6][info][application] $_GET = [ 'p' => 'admin/actions/assets/thumb' 'uid' => '5688258a-a695-4d6f-b1b1-428e8d06b8e8' 'width' => '60' 'height' => '60' 'v' => '1531837664' ]`

I had to comment on the line in the project where I have the problem.
Thanks in advance and sorry for my english.

**Comment by nokimaro** [comment]
How to check if `set_time_limit();` disabled
```
if (strpos(ini_get('disable_functions'), 'set_time_limit') === false) {
    set_time_limit(0);
}
```

**Comment by rob006** [comment]
@nokimaro It probably will not work, since `set_time_limit()` may be blocked by safe mode, not `disable_functions`. We probably need something like:

```php
@set_time_limit(0) or Yii::warning(...);
```

**Comment by Herokid** [comment]
I try with disabled_function and works correctly. 'set_time_limit' is in "disabled_function" string.
Thanks!

**Comment by angrybrad** [comment]
Pull request: https://github.com/yiisoft/yii2/pull/16536

**Comment by rob006** [comment]
Basically any function may be disabled by `disable_functions`. This creates dangerous precedent which may result in constant requests to check whether the function is available before using it.

**Comment by angrybrad** [comment]
Any function _can_ be disabled by `disable_functions`. The reality is that there are a handful of common ones that tend to be disabled in shared hosting environments.

**Comment by rob006** [comment]
The reality is that some really weird functions sometimes becomes disabled, AFAIR we already had reports about `symlink()` or `ini_set()`. `set_time_limit()` was reported once (?), request was rejected and workaround is really easy: https://github.com/yiisoft/yii2/pull/15679#issuecomment-388757423 . 
2 reports for 5 years is not really that common. 

But if we really want to fix it, I would not handle this by searching function name in `disable_functions` setting and I would not simply ignore error - it should be logged. `function_exists` or `@` operator should be faster and does not explicitly suggest that we support some exotic `disable_functions` configurations.

**Comment by angrybrad** [comment]
Fair enough... clever workaround, too.

**Comment by brandonkelly** [comment]
That workaround linked above [doesn’t work](https://www.php.net/manual/en/function.function-exists.php#67947) if `set_time_limit()` is disabled via `disable_functions`:

> function_exists will return false for functions disabled with the disable_functions ini directive. However those functions are still declared so trying to define them yourself will fail. 

So it would be worth updating Response.php to either wrap the `set_time_limit()` call with a `function_exists()` check (https://github.com/yiisoft/yii2/pull/15679#issuecomment-366760359), or remove the call altogether (https://github.com/yiisoft/yii2/pull/15679#issuecomment-366876634).

**Comment by samdark** [comment]
Wrapping it sounds OK.

**Comment by kamarton** [comment]
I like @brandonkelly solution.

Although not part of this issue, I do not think it is good idea to use `set_time_limit(0)` in web request.
